### ISSUE TITLE
Use `SizeProvider` on the components where it is applicable

### ISSUE THREAD

**Issue by spydon** [flame-engine/flame#2087]
# What could be improved
The `SizeProvider` mixin should be used on some components derived from `PositionComponent` (possibly others too).

This ticket is about investigating and adding `SizeProvider` to classes that could use it but currently don't have it.

# Why should this be improved
Since the components that don't have it can't make use of the `SizeProvider`.

# Any risks?
I don't think so, I don't think there is much that has to be changed to implement this.

**Comment by idlehart** [comment]
I should be able to figure this out 😍

**Comment by st-pasha** [comment]
The problem with having `SizeProvider` on `PositionComponent` is that many components that derive from `PositionComponent` treat their `size` property as read-only. That is, they may declare that their size is, say, 100x100, but they wouldn't work properly if some external code changed to, say, 200x50.

Some components do allow such flexibility -- and it is these components that should declare that they implement `SizeProvider`.

In other words, if you add `SizeProvider` on all `PositionComponent`s, then it would mark those components as valid targets for `SizeEffect`s. But since they can't really take those size changes into account, the result would be that the component doesn't change in size even if a size effect was applied. So, people will complain that things don't work correctly, and file bugs.

**Comment by idlehart** [comment]
So this would only work for some components and would cause massive frustration for the ones that would not accept the size?

**Comment by spydon** [comment]
> The problem with having `SizeProvider` on `PositionComponent` is that many components that derive from `PositionComponent` treat their `size` property as read-only. That is, they may declare that their size is, say, 100x100, but they wouldn't work properly if some external code changed to, say, 200x50.

The user can still change the `size` of those components, just like they can with the `SizeEffect. If the user changes the `Size` of a component that it doesn't intend to change the size on I think that should be classified as the user's fault. And also, which built-in components are you talking about that treat their `size` as read-only currently?

**Comment by st-pasha** [comment]
> And also, which built-in components are you talking about that treat their size as read-only currently?

`ClipComponent`, `IsometricTileMapComponent`, `ParticleSystemComponent`, `PolygonComponent`, etc. Generally, any component which doesn't declare it is implementing `SizeProvider`.

> The user can still change the size of those components, just like they can with the SizeEffect.

Indeed, this is an unfortunate side-effect of our API choices for the `PositionComponent`. But at least when it comes to `SizeEffect`s we've built some degree of type safety with the `SizeProvider` interface. After all, it's more likely that the user would want to apply a size effect to a component, rather than manipulate its `size` property directly. 

For example, if you have a `PolygonComponent`, and trying to apply a SizeEffect to it -- the game will currently tell you that this is not possible.

**Comment by spydon** [comment]
> > And also, which built-in components are you talking about that treat their size as read-only currently?
> 
> `ClipComponent`, `IsometricTileMapComponent`, `ParticleSystemComponent`, `PolygonComponent`, etc. Generally, any component which doesn't declare it is implementing `SizeProvider`.

`ClipComponent` could use `SizeProvider` if I read the code correctly.
But I see your point, I'll re write this ticket to be about investigating whether any components is missing it then instead.

**Comment by idlehart** [comment]
Nice, i will start looking into what components will allow the use of SizeProvider :) Good compromise.

**Comment by spydon** [comment]
Re-opening this since at least `RectangleComponent` should have the `SizeProvider` too, and probably more components. So it needs some more investigation.

**Comment by spydon** [comment]
I think all has the `SizeProvider` now.
